### ISSUE TITLE
mat: Better panic message when asking to slice to zero

### ISSUE THREAD

**Issue by btracey** [gonum/gonum#551]
The following code

```
package main

import "gonum.org/v1/gonum/mat"

func main() {
	a := mat.NewDense(5, 3, nil)
	a = a.Slice(0, 0, 0, 3).(*mat.Dense)
}
```

Gives the panic of `panic: matrix: index out of range`. A better panic would be about a zero-sized slice.

**Comment by ahlusar1989** [comment]
@btracey Is this where you would like more informed messages: https://github.com/gonum/gonum/blob/master/mat/dense.go#L49

**Comment by kortschak** [comment]
No, it would be a check for i relative to k and j relative to l [before this](https://github.com/gonum/gonum/blob/72e79ca98526cae0807f57ef69033b767f715ea7/mat/dense.go#L285) (with a change there of s/k <= i/k < i/ and s/l <= j/l < j/.

```
	if i == k || j == l {
		panic(ErrZeroLength)
	}
```

**Comment by ahlusar1989** [comment]
Gotcha. @btracey Would you like for me to take a stab at it, or do you feel that you already have a solution in mind. I am looking to learn and grow my skillset, so it will be a learning experience for me, either way :)

**Comment by btracey** [comment]
It would be great if you would want to look into this @ahlusar1989

**Comment by ahlusar1989** [comment]
@kortschak I am working on the solution for this. Do you mind clarifying for me this:

```

 s/k <= i/k < i/ and s/l <= j/l < j/

```

**Comment by kortschak** [comment]
The substitutions above are to change `k <= i` to `k < i` and the same for `l` and `j`. These are since the equality is already checked. The alternative that @btracey has suggested is to leave those the same and move the `ErrZeroLength` panic check inside the conditional below like so:
```
	mr, mc := m.Caps()
	if i < 0 || mr <= i || j < 0 || mc <= j || k < i || mr < k || l < j || mc < l {
		if i == k || j == l {
			panic(ErrZeroLength)
		}
		panic(ErrIndexOutOfRange)
	}
```

This has a minor reduction in comparisons for the fast path, so probably do that.

**Comment by kortschak** [comment]
This was fixed by #558.
